Collection checklist (#90)
ansible-collections/overview: collection_requirements.rst

*****************************

.. contents:: Topics

Overview
========

**Warning:** Subject to frequent updates
       This is a "living document", expect it to change as we progress with the Collections work over the next few months.

Feedback
========

As with any project it's very important that we get feedback from users, contributors and maintainers. We recognize that the move to Collections is a big change in how Ansible is developed and delivered.

Please raise feedback by

* Discussing in ``#ansible-community`` in Freenode IRC
* Adding to the `Community Working Group IRC meeting <https://github.com/ansible/community/issues/539>`_
* Creating `GitHub Issues <https://github.com/ansible-collections/overview/issues>`_ against this repo

Why is this needed
===================

Repo structure
===============

galaxy.yml
----------

* tags MUST set
* dependencies MUST be set to ``'>=1.0.0'`` (Might have to think about dependency order when publishing)

meta/runtime.yml
----------------
Example: `meta/runtime.yml <https://github.com/ansible-collections/collection_template/blob/main/meta/runtime.yml>`_
* MUST define the minimim version of Ansible which this collection works with

>   * If the Collection works with Ansible 2.9, then this shuld be set to

Reviewer: ```suggestion
  * If the Collection works with Ansible 2.9, then this should be set to `>=2.9.10`
```

Modules & Plugins
------------------

Documentation
~~~~~~~~~~~~~~

> All module and plugin ``DOCUMENTATION`` MUST

Reviewer: ```suggestion
All module and plugin ``DOCUMENTATION`` and ``RETURN`` MUST
```

> * Use the FQCN for ``M(...)`` FIXME Docs link

Reviewer 2: ```suggestion
* Use the FQCN for ``M(...)`` See `Linking within module documentation<https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation>`_
```

> * Use the FQCN for ``SEEALSO`` FIXME Docs link

Reviewer: ```suggestion
* Use the FQCN for ``seealso`` FIXME Docs link
```

> * Use the FQCN for ``extends_documentation_fragment:``, unles you are referring to doc_fragments from ansible-base

Reviewer 3: ```suggestion
* Use the FQCN for ``extends_documentation_fragment:``, unles you are referring to doc_fragments from ansible-base
* Use the FQCN for ``EXAMPLES`` FIXME Docs link
```

Reviewer: This should probably be after this bullet list, since that bullet list is about `DOCUMENTATION` (and not `EXAMPLES`).

Author: Good call. I've made `EXAMPLES` it's own section

> * Use the FQCN for ``extends_documentation_fragment:``, unles you are referring to doc_fragments from ansible-base

Reviewer: I wonder whether it's better to use `ansible.builtin.xxx` instead of the short name. (Does this actually work with Ansible 2.9 though?)

Contributor Workflow
====================

Changelogs
~~~~~~~~~~

> FIXME: What's the actual minimum requirement here, is it a file called X?

Reviewer: Preferred (in descending order):
1. Use antsibull-changelog;
2. Provide `changelogs/changelog.yaml` in the [correct format](https://github.com/ansible-community/antsibull-changelog/blob/main/docs/changelog.yaml-format.md);
3. Provide a link to the changelog file (self-hosted).

The big question is how that link should be provided. We probably need to create a small DB somewhere with that info.

FIXME: Add details and examples of automated system

Versioning and deprecation
~~~~~~~~~~~~~~~~~~~~~~~~~~

FIXME Details here
* Deprecation vs Collection version

> * How many Collection major versions to give

Reviewer: I'm not sure what you mean with the two points. Do we need anything else, except that they stick to semver?

Author: How many major versions do things need to stick around for before they can be deleted?

Reviewer: Hmm, good question. I think we should expect collections to give at least a ~6 months deprecation warning (i.e. one Ansible release cycle). So if they release major versions once every month, they should have a deprecation period of six major versions, but if they release only once or twice per year, a deprecation period of one major verison is fine.

(Deprecation period of one major version for me means that that there is at least one x.0.0 release where the thing is deprecated, and it is removed in (x+1).0.0. So if it is deprecated in x.0.0 it can be removed in (x+1).0.0, but if it is deprecated in x.y.z with (y,z) != (0,0) it must not be removed in (x+1).0.0 if (x+1).0.0 is released less than 6 months after x.y.z. This is satisfied by the deprecation rules for comunity.general and community.network. I'm not sure whether official Ansible content satisfies this though - the rules which were once posted somewhere didn't imply this IIRC.)

Reviewer: How about the following:
```.rst
* Every ansible minor version series (2.10.x) will keep the major version of a collection constant. So if `community.general` 1.0.0 is included in 2.10.0, 2.10.x will contain the latest `community.general` 1.y.z release that was available when building 2.10.x. If `community.general` 2.0.0 is released, it will **not** be included in ansible 2.10.x, but only in 2.11.0 (if it was published before that was released).
* Therefore, please make sure that the current major release of your collection included in 2.10.0 receives at least bugfixes as long new 2.10.x releases are produced.
* Since new minor releases are included, you can include new features, modules and plugins. You must make sure that you do not break backwards compatibility! (See [semantic versioning](https://semver.org/).) This means in particular:
  * You can fix bugs in patch releases, but not add new features or deprecate things.
  * You can add new features and deprecate things in minor releases, but not remove things or change behavior of existing features.
  * Removals and breaking changes can only be made in major releases.
* Make sure that if a deprecation is added in a collection version that is included in 2.10.x, that the removal itself will only happen in a collection version included in 2.12.0 (or later), but **not** in a collection version included in 2.11.0.
* Content moved from ansible/ansible that was scheduled for removal in 2.11 or later must NOT be removed in the current major release available when ansible 2.10.0 is released. Otherwise it would already be removed in 2.10, unexpectedly for users!
```
(Modulo better wording, that is ;) )

CC @gundalow @abadger @samccann @acozine

Reviewer 4: could we replace `2.11.0` by something more general like 'the next minor ansible following release' ?

Author: Done in https://github.com/ansible-collections/overview/pull/94/files

Repo management
===============

Repo name
~~~~~~~~~

For collections under ansible-collections the repo SHOULD be named ``NAMESPACE.COLLECTION``

Branch name and configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All new repos under `ansible-collections<https://github.com/ansible-collections>`_ MUST have `main`` as the default branch.

Existing repos SHOULD be converted to use ``main``

Repo Protections:
* Allow merge commits: disallowed

Branch protections MUST be enforced
* Require linear history
* Include administrators

CI Testing
===========

> At a minimim ``ansible-test sanity`` MUST be run from the ``ansible-base:stable-2.10`` branch

Reviewer: Maybe change this to "latest stable branch"? Or do we want an explicit list here which we have to regularly update?

For most repos GitHub actions are sufficient, see `example<https://github.com/ansible-collections/collection_template/tree/main/.github/workflows>`_

> 
> 
> 

Reviewer 3: ```suggestion

Integration Testing (if possible to implement)
==============================================

FIXME write a guide "How to write CI tests from scratch / add to existing" and put the reference here
```

Reviewer 3: or maybe `Testing (CI, Units)` instead of `CI testing` a bit above

Collections and Working Groups
==============================

> 

Reviewer 3: ```suggestion
* Working group page on a corresponding wiki (if needed. Makes sense if there is a group of modules for working with one common entity, e.g. postgresql, zabbix, grafana, etc.)
```

Reviewer 3: it's not actually a requirement, just would be nice to have for collections in community.* (~reminder/suggestion)

* Issue for agenda (or pinboard if there aren't regular meetings) as pinned issue in the Repo

> 

Reviewer 3: ```suggestion

When moving modules between collections
=======================================

All related entities must be moved / copied including:
* related plugins/module_utils/ files (moving be sure it is not used by other modules, otherwise copy)
* CI and unit tests
* corresponding documentation fragments from plugins/doc_fragments

Also:
* to change M(), examples, seealso, extended_documentation_fragments to use actual FQCNs

```

*****************************
Ansible Collections Checklist
*****************************

.. contents:: Topics

Overview
========

**Warning:** Subject to frequent updates
       This is a "living document", expect it to change as we progress with the Collections work over the next few months.

Feedback
========

As with any project it's very important that we get feedback from users, contributors and maintainers. We recognize that the move to Collections is a big change in how Ansible is developed and delivered.

Please raise feedback by

* Discussing in ``#ansible-community`` in Freenode IRC
* Adding to the `Community Working Group IRC meeting <https://github.com/ansible/community/issues/539>`_
* Creating `GitHub Issues <https://github.com/ansible-collections/overview/issues>`_ against this repo

> 

Reviewer 3: ```suggestion

Be sure you're subscribed to:
* `The issue <https://github.com/ansible-collections/overview/issues/45>` to track changes that Collection maintainers should be aware of 
* The Bullhorn, a newsletter for the Ansible developer community, <reference to where people can subscribe>
```
As @felixfontein suggested in IRC

Why is this needed
===================

Repo structure
===============

galaxy.yml
----------

* tags MUST set
* dependencies MUST be set to ``'>=1.0.0'`` (Might have to think about dependency order when publishing)

meta/runtime.yml
----------------
Example: `meta/runtime.yml <https://github.com/ansible-collections/collection_template/blob/main/meta/runtime.yml>`_
* MUST define the minimim version of Ansible which this collection works with
  * If the Collection works with Ansible 2.9, then this should be set to `>=2.9.10`

Modules & Plugins
------------------

Documentation
~~~~~~~~~~~~~~
All module and plugin ``DOCUMENTATION`` and ``RETURN`` MUST:

* Use the FQCN for ``M(...)`` See `Linking within module documentation<https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation>`_
* Use the FQCN for ``seealso`` FIXME Docs link

All module and plugin ``EXAMPLES`` MUST:

* Use FQCN for module (or plugin) name.

Other items:

> * You MUST Use the FQCN for ``extends_documentation_fragment:``, unles you are referring to doc_fragments from ansible-base

Author: @abadger @acozine What was the decision from the discussions in `#ansible-devel` regarding modules and plugins in ansible-base and how you should refer to them in docs?

Reviewer 2: FQCN everywhere.  so use `ansible.builtin.foo`

Reviewer 5: We will use FQCN everywhere, with a banner on the `ansible.builtin` modules/plugins to describe their special case. See https://github.com/ansible-community/antsibull/issues/83.

Branch name and configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All new repos under `ansible-collections<https://github.com/ansible-collections>`_ MUST have `main`` as the default branch.

Existing repos SHOULD be converted to use ``main``

Repo Protections:
* Allow merge commits: disallowed

Branch protections MUST be enforced
* Require linear history
* Include administrators

CI Testing
===========

At a minimim ``ansible-test sanity`` MUST be run from the `latest stable ansible-base branch <https://github.com/ansible/ansible/branches/all?query=stable->`_

For most repos GitHub actions are sufficient, see `example<https://github.com/ansible-collections/collection_template/tree/main/.github/workflows>`_

FIXME to write a guide "How to write CI tests" (from scratch / add to existing) and put the reference here

Unit Testing
============

Collections and Working Groups
==============================
* Working group page(s) on a corresponding wiki (if needed. Makes sense if there is a group of modules for working with one common entity, e.g. postgresql, zabbix, grafana, etc.)
* Issue for agenda (or pinboard if there aren't regular meetings) as pinned issue in the Repo

> When moving modules between collections
> =======================================

Reviewer 3: ```suggestion
When moving modules between collections
=======================================
```
After @samccann 's pr https://github.com/ansible/ansible/pull/70243 is merged, would be good to have
the reference to the fixed related part of the doc (and maybe the reference to
https://github.com/ansible-collections/overview/blob/master/README.rst#q-what-should-i-do-to-move-plugins-across-collections-during-migration)

Author: Good call, added a note on https://github.com/ansible/ansible/pull/70243#issuecomment-666480846

Preferred (in descending order):

1. Use antsibull-changelog (preffered)
2. Provide ``changelogs/changelog.yaml`` in the `correct format <https://github.com/ansible-community/antsibull-changelog/blob/main/docs/changelog.yaml-format.md>`_
3. Provide a link to the changelog file (self-hosted) (not recommended)

Versioning and deprecation
~~~~~~~~~~~~~~~~~~~~~~~~~~

FIXME Details here
* Deprecation vs Collection version
* How many Collection major versions to give

Repo management
===============

Repo name
~~~~~~~~~

For collections under ansible-collections the repo SHOULD be named ``NAMESPACE.COLLECTION``

Branch name and configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All new repos under `ansible-collections<https://github.com/ansible-collections>`_ MUST have `main`` as the default branch.

Existing repos SHOULD be converted to use ``main``

Repo Protections:
* Allow merge commits: disallowed

Branch protections MUST be enforced
* Require linear history
* Include administrators

CI Testing
===========

> At a minimim ``ansible-test sanity`` MUST be run from the `latest stable ansible-base branch <https://github.com/ansible/ansible/branches/all?query=stable->`_

Reviewer: ```suggestion
At a minimim ``ansible-test sanity`` MUST be run from the `latest stable ansible-base branch <https://github.com/ansible/ansible/branches/all?query=stable->`_. We suggest to *additionally* run ``ansible-test sanity`` from the ``devel`` branch so that you find out about new linting requirements earlier.
```

*****************************
Ansible Collections Checklist
*****************************

.. contents:: Topics

Overview
========

> 

Reviewer 4: a small intro sentence for the audience.

> This document is for maintainers of collections to provide them help, advice and guidance...

> 

Author: ```suggestion
This document is for maintainers of collections to provide them help, advice, and guidance on making sure their collections are correct.
```

**Warning:** Subject to frequent updates
       This is a "living document", expect it to change as we progress with the Collections work over the next few months.

> 

Reviewer 4: We want a small intro like
````
Goal of this document
==============

This document provides help, advice and guidance for the maintainers of collections to integration easily with Ansible.
...
````

Feedback
========

As with any project it's very important that we get feedback from users, contributors and maintainers. We recognize that the move to Collections is a big change in how Ansible is developed and delivered.

Please raise feedback by

* Discussing in ``#ansible-community`` in Freenode IRC
* Adding to the `Community Working Group IRC meeting <https://github.com/ansible/community/issues/539>`_
* Creating `GitHub Issues <https://github.com/ansible-collections/overview/issues>`_ against this repo

> Be sure you're subscribed to:

Reviewer 4: I'd put this part in a dedicated section, about how to be informed of changes and move before the feedback.

> Be sure you're subscribed to:

Author: ```suggestion
Keeping informed
================

Be sure you're subscribed to:
```

* `Changes impacting Collections <https://github.com/ansible-collections/overview/issues/45>` to track changes that Collection maintainers should be aware of 
* The Bullhorn, a newsletter for the Ansible developer community, `back issues and how to add content <https://github.com/ansible/community/issues/546>`_

Why is this needed
===================

Repo structure
===============

galaxy.yml
----------

* tags MUST be set
* dependencies MUST be set to ``'>=1.0.0'`` (Might have to think about dependency order when publishing)

meta/runtime.yml
----------------

> Example: `meta/runtime.yml <https://github.com/ansible-collections/collection_template/blob/main/meta/runtime.yml>`_

Reviewer 4: For the link perhaps it's better to use a specific commit / tag rather that using latest version as the file may evolve

> * MUST define the minimim version of Ansible which this collection works with

Reviewer: ```suggestion
* MUST define the minimum version of Ansible which this collection works with
```

>   * If the Collection works with Ansible 2.9, then this should be set to `>=2.9.10`

Reviewer: ```suggestion
  * If the collection works with Ansible 2.9, then this should be set to `>=2.9.10`
  * It's usually better to avoid adding `<2.11` as a restriction, since this for example makes it impossible to use the collection with the current ansible-base devel branch (which has version 2.11.0.dev0)
```

Modules & Plugins
------------------

Documentation
~~~~~~~~~~~~~~
All module and plugin ``DOCUMENTATION`` and ``RETURN`` MUST:

* Use the FQCN for ``M(...)`` See `Linking within module documentation<https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation>`_
* Use collection version numbers for `version_added` (otherwise `version_added_collection` must be provided, but this is discouraged)

All module and plugin ``EXAMPLES`` MUST:

* Use FQCN for module (or plugin) name.
* for modules (or plugins) left in ansible-base use ``ansible.builtin.template``

Other items:
* You MUST Use the FQCN for ``extends_documentation_fragment:``, unless you are referring to doc_fragments from ansible-base

> 

Reviewer 2: ```suggestion
* You MUST use your collection version that is included in the 2.10 Ansible package for the `version_added` field for anything added in 2.10. This field reflects the collection version now, not the ansible-base version.
```

Reviewer: That's not entirely true. You need to use the version number of the collection where the plugin / module / option was added, which is <= to the version of the collection inluded in 2.10.0. Depending on how many versoins of the collection have been released before ansible 2.10.0 will be released, there can be several different values. (In c.g, both the values 0.2.0 and 1.0.0 will appear, and maybe 1.1.0 if that's released in time to make it into 2.10.0.)

Also see https://github.com/ansible-collections/overview/pull/90/files#diff-31dcce2efaaf88ba4f5a7824b787cb9eR55 which already mentions `version_added`.

Reviewer 2: oh hah didn't see that! Will delete my comment.

Contributor Workflow
====================

Changelogs
~~~~~~~~~~

To give a consistent feel for changelogs across collections, and ensure for collections included in the ``ansible`` package we suggest you use `antsibull-changelog <https://github.com/ansible-community/antsibull-changelog>`_

Preferred (in descending order):

> 1. Use antsibull-changelog (preffered)

Reviewer 4: preferred

> 1. Use antsibull-changelog (preffered)

Author: ```suggestion
1. Use antsibull-changelog (preferred)
```

2. Provide ``changelogs/changelog.yaml`` in the `correct format <https://github.com/ansible-community/antsibull-changelog/blob/main/docs/changelog.yaml-format.md>`_
3. Provide a link to the changelog file (self-hosted) (not recommended)

> Please note that the porting guide is complied from `changelogs/changelog.yaml` (sections `breaking_changes`, `major_changes`, `deprecated_features`, `removed_features`). So if you use option 3, you will not be able to add something to the porting guide.

Reviewer: ```suggestion
Please note that the porting guide is complied from ``changelogs/changelog.yaml`` (sections ``breaking_changes``, ``major_changes``, ``deprecated_features``, ``removed_features``). So if you use option 3, you will not be able to add something to the porting guide.
```
Forgot this is RST :)

> Please note that the porting guide is complied from `changelogs/changelog.yaml` (sections `breaking_changes`, `major_changes`, `deprecated_features`, `removed_features`). So if you use option 3, you will not be able to add something to the porting guide.

Reviewer 4: compiled ?

Reviewer 4: I seen `minor_changes` in ansible, is it also in collections ?

Reviewer: Yes, that's the main category for new features / options.

Versioning and deprecation
~~~~~~~~~~~~~~~~~~~~~~~~~~

FIXME Details here
* Deprecation vs Collection version
* How many Collection major versions to give

Repo management
===============

Repo name
~~~~~~~~~

For collections under ansible-collections the repo SHOULD be named ``NAMESPACE.COLLECTION``

Branch name and configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

All new repos under `ansible-collections<https://github.com/ansible-collections>`_ MUST have `main`` as the default branch.

Existing repos SHOULD be converted to use ``main``

Repo Protections:
* Allow merge commits: disallowed

Branch protections MUST be enforced
* Require linear history
* Include administrators

CI Testing
===========

> At a minimim ``ansible-test sanity`` MUST be run from the `latest stable ansible-base branch <https://github.com/ansible/ansible/branches/all?query=stable->`_. We suggest to *additionally* run ``ansible-test sanity`` from the ``devel`` branch so that you find out about new linting requirements earlier.

Reviewer: ```suggestion
At a minimum ``ansible-test sanity`` MUST be run from the `latest stable ansible-base branch <https://github.com/ansible/ansible/branches/all?query=stable->`_. We suggest to *additionally* run ``ansible-test sanity`` from the ansible/ansible ``devel`` branch so that you find out about new linting requirements earlier.
```

**Warning:** Subject to frequent updates
       This is a "living document", expect it to change as we progress with the Collections work over the next few months.

Feedback
========

As with any project it's very important that we get feedback from users, contributors and maintainers. We recognize that the move to Collections is a big change in how Ansible is developed and delivered.

Please raise feedback by

* Discussing in ``#ansible-community`` in Freenode IRC
* Adding to the `Community Working Group IRC meeting <https://github.com/ansible/community/issues/539>`_
* Creating `GitHub Issues <https://github.com/ansible-collections/overview/issues>`_ against this repo

Keeping informed
================

Be sure you're subscribed to:
* `Changes impacting Collections <https://github.com/ansible-collections/overview/issues/45>` to track changes that Collection maintainers should be aware of 
* The Bullhorn, a newsletter for the Ansible developer community, `back issues and how to add content <https://github.com/ansible/community/issues/546>`_

Why is this needed
===================

> Repo structure

Reviewer 6: Better avoid shortened name, so Repository.
sorry to nit-pick

Author: Good call, fixed all.

===============

galaxy.yml
----------

* tags MUST be set
* dependencies MUST be set to ``'>=1.0.0'`` (Might have to think about dependency order when publishing)

meta/runtime.yml
----------------
Example: `meta/runtime.yml <https://github.com/ansible-collections/collection_template/blob/main/meta/runtime.yml>`_
* MUST define the minimum version of Ansible which this collection works with
  * If the collection works with Ansible 2.9, then this should be set to `>=2.9.10`
  * It's usually better to avoid adding `<2.11` as a restriction, since this for example makes it impossible to use the collection with the current ansible-base devel branch (which has version 2.11.0.dev0)

Modules & Plugins
------------------

Documentation
~~~~~~~~~~~~~~
All module and plugin ``DOCUMENTATION`` and ``RETURN`` MUST:

> * Use the FQCN for ``M(...)`` See `Linking within module documentation<https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation>`_

Reviewer 3: ```suggestion
* Use the FQCN for ``M(...)`` and ``- module:`` references of ``seealso`` subsections. See `Linking within module documentation<https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation>`_
```

* Use collection version numbers for `version_added` (otherwise `version_added_collection` must be provided, but this is discouraged)

All module and plugin ``EXAMPLES`` MUST:

* Use FQCN for module (or plugin) name.

> * for modules (or plugins) left in ansible-base use ``ansible.builtin.template``

Reviewer 3: ```suggestion
* For modules (or plugins) left in ansible-base use ``ansible.builtin.`` as a FQCN prefix, for example, ``ansible.builtin.template``
```

Other items:
* You MUST Use the FQCN for ``extends_documentation_fragment:``, unless you are referring to doc_fragments from ansible-base

Contributor Workflow
====================

Changelogs
~~~~~~~~~~

To give a consistent feel for changelogs across collections, and ensure for collections included in the ``ansible`` package we suggest you use `antsibull-changelog <https://github.com/ansible-community/antsibull-changelog>`_

Preferred (in descending order):

1. Use antsibull-changelog (preferred)
2. Provide ``changelogs/changelog.yaml`` in the `correct format <https://github.com/ansible-community/antsibull-changelog/blob/main/docs/changelog.yaml-format.md>`_
3. Provide a link to the changelog file (self-hosted) (not recommended)

Please note that the porting guide is compiled from ``changelogs/changelog.yaml`` (sections ``breaking_changes``, ``major_changes``, ``deprecated_features``, ``removed_features``). So if you use option 3, you will not be able to add something to the porting guide.

Versioning and deprecation
~~~~~~~~~~~~~~~~~~~~~~~~~~

FIXME Details here
* Deprecation vs Collection version
* How many Collection major versions to give

> Repo management

Reviewer 6: same for use of "Repository" over "Repo"

===============

> Repo name

Reviewer 3: ```suggestion
Repo name
```
Roaming around the collections i noticed a bit strange things as parts of FQCNs like `vyos.vyos.*`. We definitely should put some recommendations here. Maybe later though

Reviewer 6: For documentation, avoid shortened name also, if we can use "Repository"

Reviewer 6: Yes, indeed, and with al modules name starting with also `vyos_` so you end up with `vyos.vyos.vyos_foo` ... . I tried to open issues about that on some projects, but now it's a bit too late.

Reviewer 6: @Andersson007 Do you think we should providing some naming advice for the collections ?

Reviewer 3: @bmillemathias sure, it could be general points without giving concrete examples how people shouldn't do :)
E.g. meaning (and we can see something similar in every book about programming)
```
- try not to use very long names but if you think it's better / necessary, feel free to do it
- choosing a name try to avoid repetitive parts in it
- ...
```
Something like that.
I definitely shoul think about it. If you feel the same and have something to suggest, it would be really great:)
